src/nrnoc/multicore.cpp

```diff
@@ -391,24 +391,23 @@ void nrn_threads_free() {
         for (tml = nt->tml; tml; tml = tml2) {
             Memb_list* ml = tml->ml;
             tml2 = tml->next;
-            free((char*) ml->nodelist);
-            free((char*) ml->nodeindices);
-            delete[] ml->prop;
+            free((char*) std::exchange(ml->nodelist, nullptr));
+            free((char*) std::exchange(ml->nodeindices, nullptr));
+            delete[] std::exchange(ml->prop, nullptr);
             if (!memb_func[tml->index].hoc_mech) {
-                free((char*) ml->pdata);
+                free((char*) std::exchange(ml->pdata, nullptr));
             }
             if (ml->_thread) {
                 if (memb_func[tml->index].thread_cleanup_) {
                     (*memb_func[tml->index].thread_cleanup_)(ml->_thread);
                 }
-                delete[] ml->_thread;
+                delete[] std::exchange(ml->_thread, nullptr);
             }
-            delete ml;
-            free((char*) tml);
+            delete std::exchange(ml, nullptr);
+            free((char*) std::exchange(tml, nullptr));
         }
```

Comment:
In terms of best practices then `free(std::exchange(...))` is awesome since we retaining full back compatibility and completely avoids double free (unless object copies were made). It exists already in a lot of places, introduced by Olli.
So while this PR doesn't depend on it (it was an attempt to find the bug) and I can of course revert, IMO we should even target the whole source code to have this before heavier refactorings (maybe except in performance critical kernels).

Reply:
Yes, completely understand rational! 

I was just thinking about this scenario where:

```
free((char*) std::exchange(ml->nodelist, nullptr));
free((char*) std::exchange(ml->nodeindices, nullptr));
...
delete std::exchange(ml, nullptr);
```
As we are deleting `ml` and setting it to `nullptr` immediately in this function, I guess we don't get any benefit from the pattern here.